File: castgeo/scala/castGeo.scala
package rwt.castgeo 

object CastGeo {
  private val line = Array("*   *", "  *  ")

  private val rnd = new java.util.Random()
  private def randomFig() : Array[String] = Array.fill(4){ line(rnd.nextInt(2)) } 

  type FigRow = Array[Array[String]]   

  // display a row of geomantic figures, with initial and middle spaces
  private def disp(initSp: Int, midSp: Int, figs : FigRow) : Unit = 
     figs.transpose.foreach { row =>
         print( row.mkString(" " * initSp, " " * midSp, "\n") )
     }

  // combine adjacent geomantic figures from a row, returning the resulting row
  private def combined(figs : FigRow) : FigRow =
     figs.grouped(2).map{ pair => 
       (pair(0), pair(1)).zipped.map { (a,b) => line(if (a==b) 0 else 1) }
     }.toArray

  def main(args: Array[String]) : Unit = {
     val moms      = Array.fill(4)(randomFig)
     val daughters = moms.transpose
     val line1     = (moms ++ daughters).reverse
     val nieces    = combined(line1)
     val witnesses = combined(nieces)
     val judge     = combined(witnesses)
     disp(2,5,line1)        ; println()
     disp(7,15,nieces)      ; println()
     disp(17,35,witnesses)  ; println()
     disp(37,0,judge) 
  }
}

File: castgeo/haskell/castgeo.hs
-- cast a geomantic shield
import Data.List(transpose, intersperse)
import qualified System.Random as Rnd

type Figure = [Bool]

dispFigs :: Int -> Int -> [Figure] -> [String]  -- display a line of figures
dispFigs sp1 spMid figs = map comb strs 
  where strs     = map (map toStr) (transpose figs)
        comb ln  = concat $ (spaces sp1) : (intersperse (spaces spMid) ln)
        spaces n = take n (repeat ' ')
        toStr bl = if bl then "  *  " else "*   *"

combine :: [Figure] -> [Figure]  -- combine adjacent figures 
combine [] = []
combine (a : b : fs) = (mergeFigs a b) : (combine fs)
 where mergeFigs = zipWith (\a b -> if a == b then False else True) 

rfig :: IO Figure
rfig = sequence $ take 4 $ repeat $ Rnd.getStdRandom (Rnd.random)
   
displine l = (mapM_ putStrLn l) >> (putStrLn "")

main = do
  moms <- sequence $ take 4 (repeat rfig)
  let line1 = reverse (moms ++ (transpose moms))
  let nieces = combine line1 
  let witnesses = combine nieces
  let judge = combine witnesses
  displine $ dispFigs 2 5 line1 
  displine $ dispFigs 7 15 nieces
  displine $ dispFigs 17 35 witnesses
  displine $ dispFigs 37 0 judge
